contracts/staking/StakingMessages.sol

```diff
@@ -0,0 +1,289 @@
+        internal
+        pure
+        returns (bytes32, bytes memory)
+    {
+        (bytes32 validationID, bytes memory serializedValidationInfo) =
+            packValidationInfo(valiationInfo);
+
+        bytes memory res = new bytes(148);
+        // Pack the message type
+        for (uint256 i; i < 4; ++i) {
+            res[i] = bytes1(uint8(SUBNET_VALIDATOR_REGISTRATION_MESSAGE_TYPE_ID >> (8 * (3 - i))));
+        }
+        // Pack the validation info
+        for (uint256 i; i < 144; ++i) {
+            res[i + 4] = serializedValidationInfo[i];
+        }
+
+        return (validationID, res);
+    }
+
+    /**
+     * @notice Unpacks a byte array as a RegisterSubnetValidator message.
+     * The message format specification is:
+     * +--------------+----------+----------+
+     * |       typeID :   uint32 |  4 bytes |
+     * +--------------+----------+----------+
+     * | validationID : [32]byte | 32 bytes |
+     * +--------------+----------+----------+
+     * |        valid :     bool +  1 byte  |
+     * +--------------+----------+----------+
+     *                           | 37 bytes |
+     *                           +----------+
+     * @return The validationID and whether or the validation period was registered
+     * or is not a validator and never will be a validator to do the expiry time passing.
+     */
+    function unpackSubnetValidatorRegistrationMessage(bytes memory input)
+        internal
+        pure
+        returns (bytes32, bool)
+    {
+        require(input.length == 37, "StakingMessages: Invalid message length");
+
+        // Unpack the type ID
+        uint32 typeID;
+        for (uint256 i; i < 4; ++i) {
+            typeID |= uint32(uint8(input[i])) << uint32((8 * (3 - i)));
+        }
+        require(
+            typeID == SUBNET_VALIDATOR_REGISTRATION_MESSAGE_TYPE_ID,
+            "StakingMessages: Invalid message type"
+        );
+
+        // Unpack the validation ID.
+        bytes32 validationID;
+        for (uint256 i; i < 32; ++i) {
+            validationID |= bytes32(uint256(uint8(input[i + 4])) << (8 * (31 - i)));
+        }
+
+        // Unpack the validity
+        bool valid = input[36] != 0;
+
+        return (validationID, valid);
+    }
+
+    /**
+     * @notice Packs a SetSubnetValidatorWeight message into a byte array.
+     * The message format specification is:
+     * +--------------+----------+----------+
+     * |       typeID :   uint32 |  4 bytes |
+     * +--------------+----------+----------+
+     * | validationID : [32]byte | 32 bytes |
+     * +--------------+----------+----------+
+     * |        nonce :   uint64 |  8 bytes |
+     * +--------------+----------+----------+
+     * |       weight :   uint64 |  8 bytes |
+     * +--------------+----------+----------+
+     *                           | 52 bytes |
+     *                           +----------+
+     */
+    function packSetSubnetValidatorWeightMessage(
```

Comment:
This can be simplified to:

```solidity
return abi.encodePacked(
    SET_SUBNET_VALIDATOR_WEIGHT_MESSAGE_TYPE_ID,
    validationID,
    nonce,
    weight
);
```

Reply:
There are similar improvements that can be made to unpacking, but it was easier to write code so I created [a PR, based off this branch, to demonstrate](https://github.com/ava-labs/teleporter/pull/482).

Together they save about 39,500 gas on the pack-unpack round trip.

Reply:
I only just saw the `TODO` re gas efficiency 🤦 sorry!